libs/error-stack/src/fmt.rs

```diff
@@ -133,6 +133,148 @@
 //! </pre>
 //!
+//! ## Implementation Details
+//!
+//! Nothing explained here is under any semver guarantee. This section explains the algorithm used
+//! to produce the [`Debug`] output of a [`Report`].
+//!
+//! During the explanation we will make use of two different [`Report`]s, the overview tree (shown
+//! first) only visualizes contexts, while the second, more detailed tree shows attachments and
+//! contexts.
+//!
+//! In the detailed tree the type of [`Frame`] is distinguished using a superscript letter, `ᵃ` is
+//! used to indicate attachments and `ᶜ` is used to indicate contexts. For clarity the overview tree
+//! uses digits, while the detailed tree uses letters for different [`Frame`]s.
+//!
+//! Overview (Context only) Tree:
+//!
+//! ```text
+//!     0
+//!     |
+//!     1
+//!    / \
+//!   2   6
+//!  / \  |
+//! 3   4 7
+//!     | |
+//!     5 8
+//! ```
+//!
+//!
+//! Detailed (Context + Attachment) Tree:
+//!
+//! ```text
+//!    Aᶜ
+//!    |
+//!    Bᵃ
+//!   / \
+//!  Cᵃ  Eᵃ
+//!  |   |
+//!  Dᶜ  Fᵃ
+//!     / \
+//!    Gᵃ  Iᶜ
+//!    |
+//!    Hᶜ
+//! ```
+//!
+//! During formatting we distinguish between two cases (for contexts):
+//!
+//! * Lists
+//! * Groups
+//!
+//! in this explanation lists are delimited by `[` and `]`, while groups are delimited by `(` and
+//! `)`.
+//!
+//! While formatting we view the [`Report`]s as a tree of [`Frame`]s, therefore the following
+//! explanation will use terminology associated with trees, every [`Frame`] is a node and can have
+//! `0..n` children, a node that has no children (a leaf) is guaranteed to be a [`Context`].
+//!
+//! A list is a list of nodes where each node in the list if the parent of the following element and
+//! has only a single child, the last element of a list, can have `0..n` children. In the examples
+//! above, `[6, 7, 8]` is considered a list, while `[1, 6]` is not, because while `1` is a parent of
+//! `6`, `1` has more than 1 child.
+//!
+//! A group is a list of nodes where each node shares a common immediate context parent that has
+//! more than `1` child, this means that `(2, 6)` is a group (they share `1` as an immediate context
+//! parent), while `(3, 4, 6)` is not. `(3, 4, 6)` share the same parent with more than 1 child
+//! (`1`), but `1` is not the immediate context parent of `3` and `4` (`2`) is. In the more detailed
+//! example `(Dᶜ, Hᶜ, Cᶜ)` is considered a group because they share the same *immediate* context
+//! parent `Aᶜ`, important to note is that we only refer to immediate context parents, `Fᵃ` is the
+//! immediate parent of `Iᶜ`, but is not a [`Context`], therefore to find the immediate context
+//! parent, we travel up the tree until we encounter our first [`Context`] node. Groups always
+//! contain lists, for the sake of clarity this explanation only shows the first element.
+//!
+//! Additional rules that apply:
+//! * lists are never empty
+//! * lists are nested in groups
+//! * groups are always preceded by lists
```

Comment:
What does this mean? Every context is in exactly one list?

Reply:
hopefully clarified in [3a68ec5](https://github.com/hashintel/hash/pull/2156/commits/3a68ec5e73d3dd9fc8f3fa7a3777f62e2852bdde), it just means that there's always at least a single context before a group, as otherwise, a group can never happen. The only place where this might not be the case is the top level, where we render them out as two separate trees.

Reply:
additional clarification in [288470f](https://github.com/hashintel/hash/pull/2156/commits/288470f80eba6aab6bb3b3ffdb3e6ffafa89eee7)